### PR TITLE
[TensorDemux] Support other/tensors on src pads

### PR DESCRIPTION
**Pull request by gichan-jang**
Changed to suppot ohter/tensors on src pads.
Re-use of the tensor and non-sequential combinations are also possible.

Related issue: #2844    

Signed-off-by: gichan <gichan2.jang@samsung.com>

### CODE REVIEW

File: gst/nnstreamer/tensor_demux/gsttensordemux.c [hunk 1]
```diff
@@ -44,13 +44,13 @@
...
- * multifilesrc location="testsequence01_%1d.png" index=0 caps="image/png, framerate=(fraction)30/1" ! pngdec ! tensor_converter ! mux.sink_0 \
- * multifilesrc location="testsequence01_%1d.png" index=0 caps="image/png, framerate=(fraction)30/1" ! pngdec ! tensor_converter ! mux.sink_1 \
- * multifilesrc location="testsequence01_%1d.png" index=0 caps="image/png, framerate=(fraction)30/1" ! pngdec ! tensor_converter ! mux.sink_2 \
- * demux.src_0 ! queue ! filesink location=demux00.log \
- * demux.src_1 ! queue ! filesink location=demux01.log \
- * demux.src_2 ! queue ! filesink location=demux02.log
+ * gst-launch-1.0 tensor_mux name=mux ! tensor_demux name=demux tensorpick=0,1:2,2+0 \
+ * ... (tensor 0) ! mux.sink_0 \
+ * ... (tensor 1) ! mux.sink_0 \
+ * ... (tensor 2) ! mux.sink_0 \
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_demux/gsttensordemux.c:50]: sink_0 --> sink_1/2 :)

File: gst/nnstreamer/tensor_demux/gsttensordemux.c [hunk 2]
```diff
@@ -375,10 +394,40 @@ gst_tensor_demux_get_tensor_pad (GstTensorDemux * tensor_demux,
   g_free (stream_id);
   gst_event_unref (event);
 
-  caps = gst_tensor_caps_from_config (&config);
-  gst_pad_set_caps (pad, caps);
-  gst_element_add_pad (GST_ELEMENT_CAST (tensor_demux), pad);
+  if (tensor_demux->tensorpick != NULL) {
+    gchar *seleted_tensor =
+        (gchar *) g_list_nth_data (tensor_demux->tensorpick, nth);
```
Comment (dongju-chae) [inline gst/nnstreamer/tensor_demux/gsttensordemux.c:399]: Is it ensured that `tensorpick` always has `nth` data?

Comment (gichan-jang) [reply gst/nnstreamer/tensor_demux/gsttensordemux.c:399]: Right. I added error handling code (Line 403).

**Review by dongju-chae — APPROVED** [review]
LGTM

File: gst/nnstreamer/tensor_demux/gsttensordemux.c [hunk 3]
```diff
@@ -28,8 +28,8 @@
 /**
  * SECTION:element-tensor_demux
  *
- * A Deuxer that demux tensors stream to tensor stream for NN frameworks.
- * The outputs are always in the format of other/tensor
+ * A Deuxer that demux tensors stream to tensor(s) stream for NN frameworks.
```
Comment (wooksong) [inline gst/nnstreamer/tensor_demux/gsttensordemux.c:31]: Deuxer? Typo?

File: gst/nnstreamer/tensor_demux/gsttensordemux.c [hunk 4]
```diff
@@ -439,31 +495,39 @@ gst_tensor_demux_chain (GstPad * pad, GstObject * parent, GstBuffer * buf)
   /* supposed n memory blocks in buffer */
   g_assert (gst_buffer_n_memory (buf) == num_tensors);
 
-  for (i = 0; i < num_tensors; i++) {
+  if (tensor_demux->tensorpick != NULL) {
+    num_srcs = g_list_length (tensor_demux->tensorpick);
+    list = tensor_demux->tensorpick;
+  } else {
+    num_srcs = num_tensors;
+  }
```
Comment (wooksong) [inline gst/nnstreamer/tensor_demux/gsttensordemux.c:503]: How about this?
```suggestion
  num_srcs = num_tensors;
  if (tensor_demux->tensorpick != NULL) {
    num_srcs = g_list_length (tensor_demux->tensorpick);
    list = tensor_demux->tensorpick;
  }
```

File: gst/nnstreamer/tensor_demux/gsttensordemux.c (hunk 3 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_demux/gsttensordemux.c:31]: Updated!

File: gst/nnstreamer/tensor_demux/gsttensordemux.c (hunk 4 above)
Comment (gichan-jang) [reply gst/nnstreamer/tensor_demux/gsttensordemux.c:503]: Thank you for the suggestion. Updated.

**Review by myungjoo — APPROVED** [review]
LGTM. It does not incur memcpies, right?

**Comment by gichan-jang** [comment]
> LGTM. It does not incur memcpies, right?

Yes, It just adjusts the ref count of the memory.

**Review by jaeyun-jung — APPROVED** [review]

File: gst/nnstreamer/tensor_demux/gsttensordemux.c [hunk 5]
```diff
@@ -375,10 +394,46 @@ gst_tensor_demux_get_tensor_pad (GstTensorDemux * tensor_demux,
...
+    GstCaps *peer_caps;
+
+    g_assert (g_list_length (tensor_demux->tensorpick) >= nth);
+
+    seleted_tensor = (gchar *) g_list_nth_data (tensor_demux->tensorpick, nth);
+    strv = g_strsplit_set (seleted_tensor, ":+", -1);
+    num = g_strv_length (strv);
+    peer_caps = gst_pad_peer_query_caps (pad, NULL);
+
+    if (num == 1 && gst_tensor_demux_is_tensor_caps (peer_caps)) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_demux/gsttensordemux.c:410]: later work, let's consider adding common function to get caps (filter / if / demux)

**Pull request merged**